### PR TITLE
Modify rule S3039: Improve S3039 to comply to LayC

### PR DESCRIPTION
**Pull request by johann-beleites-sonarsource**

### CODE REVIEW

**Comment by johann-beleites-sonarsource** [comment]
Before:
https://next.sonarqube.com/sonarqube/coding_rules?q=s3039&open=java%3AS3039
After:
image
image

File: rules/S3039/java/rule.adoc [hunk 1]
```diff
@@ -1,32 +1,47 @@
...
+There are various `String` operations that take one or more character indexes as input and return a result that is the function of the original string and the indexes.
+Indexing in this context is zero-based, meaning that the first character's index is `0`.
+As a result, given a string `myString`, its last character is at index `myString.length - 1`.
 
+If an index is smaller than 0 or larger than `myString.length - 1` is passed to one of the string operation methods a `StringIndexOutOfBoundsException` will be thrown.
+For instance, it is not possible to use `String::substring` to reverse the order of characters in a string and it is not possible to use `String::charAt` to retrieve a value before the start or after the end of a string.
 
-This rule detects when negative literal or ``++String::length++`` is passed as an argument to the ``++String::substring++``, ``++String::charAt++`` and related methods.
+This rule raises an issue when a negative literal or `String::length` is passed as an argument to the `String::substring`, `String::charAt`, and related methods.
```
Comment (alban-auzeill) [inline rules/S3039/java/rule.adoc:10]: It depends, for `charAt ` the rule raises when `index > length() -1` and for  `substring` when `index > length()`
```suggestion
This rule raises an issue when a negative literal or a too-large index is passed as an argument to the `String::substring`, `String::charAt`, and related methods.
```

File: rules/S3039/java/rule.adoc [hunk 2]
```diff
@@ -1,32 +1,47 @@
...
+It also raises an issue when the start index passed to `String::substring` is larger than the end index.
 
 
-=== Noncompliant code example
+== How to fix it
 
-[source,java]
+Use non-negative indexes that are strictly smaller than the length of the string in question.
```
Comment (alban-auzeill) [inline rules/S3039/java/rule.adoc:16]:
```suggestion
Use non-negative indexes that are smaller than or equal to the length of the string in question with `String::substring` and strictly smaller with `String::charAt`.
```

File: rules/S3039/java/rule.adoc [hunk 3]
```diff
@@ -1,32 +1,47 @@
...
-String speech = "Now is the time for all good people to come to the aid of their country.";
+String speech = "Lorem ipsum dolor sit amet";
 
-String substr1 = speech.substring(-1, speech.length());  // Noncompliant; start and end values both bad
-String substr2 = speech.substring(speech.length(), 0); // Noncompliant, start value must be < end value
-char ch = speech.charAt(speech.length());  // Noncompliant
+String substr1a = speech.substring(0, speech.length());      // Noncompliant, speech.length() is out of bounds
+String substr1b = speech.substring(-1, speech.length() - 1); // Noncompliant, -1 is out of bounds
+String substr2 = speech.substring(speech.length - 1, 0);     // Noncompliant, the start index must be smaller than the end index. String::substring to reverse the order of characters.
+char ch = speech.charAt(speech.length());                    // Noncompliant, speech.length() is out of bounds
```
Comment (alban-auzeill) [inline rules/S3039/java/rule.adoc:31]: `substring(0, speech.length())` is a valid compliant.
`substring(speech.length - 1, 0)` is not supported by the rule, only `substring(speech.length(), 0)`
```suggestion
String substr1 = speech.substring(-1, speech.length()); // Noncompliant, -1 is out of bounds
String substr2 = speech.substring(speech.length(), 0);     // Noncompliant, the beginIndex must be smaller than or equal to the endIndex
char ch = speech.charAt(speech.length());                    // Noncompliant, speech.length() is out of bounds
```

File: rules/S3039/java/rule.adoc [hunk 4]
```diff
@@ -1,32 +1,47 @@
...
-String speech = "Now is the time for all good people to come to the aid of their country.";
+String speech = "Lorem ipsum dolor sit amet";
 
-String substr1 = speech; // Closest correct values to original code yield whole string
-String substr2 = new StringBuilder(speech).reverse().toString()
-char ch = speech.charAt(speech.length()-1);
+String substr1a = speech;                                       // Compliant, no string operation used
+String substr1b = speech.substring(0, speech.length() - 1);     // Compliant, both 0 and speech.length - 1 are in bounds
+String substr2 = new StringBuilder(speech).reverse().toString() // Compliant, the string can be reversed using StringBuilder::reverse
+char ch = speech.charAt(speech.length() - 1);                   // Compliant, speech.length() - 1 is in bounds.
```
Comment (alban-auzeill) [inline rules/S3039/java/rule.adoc:44]:
```suggestion
String substr1 = speech;                                       // Compliant, no string operation used
String substr2 = new StringBuilder(speech).reverse().toString(); // Compliant, the string can be reversed using StringBuilder::reverse()
char ch = speech.charAt(speech.length() - 1);                   // Compliant, speech.length() - 1 is in bounds.
```

**Review by alban-auzeill — CHANGES_REQUESTED** [review]

**Review by alban-auzeill — APPROVED** [review]

**Pull request merged**